pkg/kubelet/volumemanager/volume_manager.go

```diff
@@ -357,21 +357,38 @@ func (vm *volumeManager) WaitForAttachAndMount(pod *v1.Pod) error {
 		// Timeout expired
 		unmountedVolumes :=
 			vm.getUnmountedVolumes(uniquePodName, expectedVolumes)
+		// Also get unattached volumes for error message
+		unattachedVolumes :=
+			vm.getUnattachedVolumes(expectedVolumes)
+
 		if len(unmountedVolumes) == 0 {
 			return nil
 		}
 
 		return fmt.Errorf(
-			"timeout expired waiting for volumes to attach/mount for pod %q/%q. list of unattached/unmounted volumes=%v",
+			"timeout expired waiting for volumes to attach or mount for pod %q/%q. list of unmounted volumes=%v. list of unattached volumes=%v",
 			pod.Namespace,
 			pod.Name,
-			unmountedVolumes)
+			unmountedVolumes,
+			unattachedVolumes)
 	}
 
 	glog.V(3).Infof("All volumes are attached and mounted for pod %q", format.Pod(pod))
 	return nil
 }
 
+// getUnattachedVolumes returns a list of the volumes that are expected to be attached but
+// are not currently attached to the node
+func (vm *volumeManager) getUnattachedVolumes(expectedVolumes []string) []string {
+	unattachedVolumes := []string{}
+	for _, volume := range expectedVolumes {
+		if !vm.actualStateOfWorld.VolumeExists(v1.UniqueVolumeName(volume)) {
```

Comment:
I don't think this will give you the list of unattached volumes. Volumes should only be added to the volumeManager's `actualStateOfWorld` if they are mounted.

Reply:
// VolumeExists returns true if the given volume exists in the list of
// attached volumes in the cache, indicating the volume is attached to this
// node.

This was the comment on "VolumeExists," maybe I am misunderstanding it but I thought this means it does not have to be mounted to be recorded as attached in actualStateOfWorld. If not, we will have to change the documentation or make it clearer.

Reply:
when reconciler in kubelet tries to VerifyControllerAttachedVolume(), it the volume is already attached (indicated by node status's volumesattached list), it will mark volumes as attached in the actual state, so I think it is ok to use this way to get unattached volumes.

Reply:
You're right! Forgot the code I wrote :)